### ISSUE TITLE
 Occasional AccessDeniedErrors after applying ACL metadata to a soft-deleted stream

### ISSUE THREAD

**Issue by jonasholtkamp** [EventStore/EventStore#1895]
This is cross-post from [nicdex/node-eventstore-client!65](https://github.com/nicdex/node-eventstore-client/issues/65). The maintainer asked me to post this issue here because the assumption is it's rather a general Eventstore issue than a library issue.

---

Lately I've observed a strange asynchronous behavior of `setStreamMetadataRaw`. Let me explain my test setup first, then I'll go on and explain the problem.

### Prerequisites
* `node-eventstore-client`: 0.2.6
* Eventstore: 4.1.1 (running in Docker 18.09.1) with default configuration apart from a custom default ACL (see below)
* Node.js: 11.9.0
* OS: macOS 10.14.3

### Eventstore default ACL
```
  '$userStreamAcl': {
      '$r': '$admins',
      '$w': '$admins',
      '$d': '$admins',
      '$mr': '$all',
      '$mw': '$all'
    },
    '$systemStreamAcl': {
      '$r': '$admins',
      '$w': '$admins',
      '$d': '$admins',
      '$mr': '$admins',
      '$mw': '$admins'
    }
```
With that the user is not by default allowed to write to/read from any stream.

### Test setup
* for example code see here: https://gist.github.com/jonasholtkamp/a1909ecedad5f16ec4dbbe9f0c8007da
* create an ES user with exhaustive group privileges (`[ 'test-r', 'test-w', 'test-d', 'test-mr', 'test-mw' ]`)
* connect to the ES with this user
* set ACLs on an arbitrary stream (`test-1`) that match the group privileges above (this gives our user all privileges on this stream)
* append an arbitrary event to `test-1`
* **soft**-delete `test-1`
* try to read from `test-1`
* this (as expected) fails with an `AccessDeniedError` because the ACLs have been deleted as well
* set the exact same ACLs like above _again_ on `test-1`
* try to read from `test-1`
* expectation: reading succeeds for `test-1`, it's however empty

### Problem
* in some cases (<10%) the **last** reading step fails with this error:
```
(node:30306) UnhandledPromiseRejectionWarning: AccessDeniedError: Read access denied for stream 'test-1'.
    at ReadStreamEventsBackwardOperation.module.exports.ReadStreamEventsBackwardOperation._inspectResponse (/Users/[redacted]/node_modules/node-eventstore-client/lib/dist.js:5409:17)
    at ReadStreamEventsBackwardOperation.module.exports.OperationBase.inspectPackage (/Users/[redacted]/node_modules/node-eventstore-client/lib/dist.js:549:19)
    at EventStoreConnectionLogicHandler.module.exports.EventStoreConnectionLogicHandler._handleTcpPackage (/Users/[redacted]/node_modules/node-eventstore-client/lib/dist.js:3548:38)
    at /Users/[redacted]/node_modules/node-eventstore-client/lib/dist.js:3098:10
    at Immediate._handlers.(anonymous function) (/Users/[redacted]/node_modules/node-eventstore-client/lib/dist.js:3748:7)
    at processImmediate (timers.js:639:17)
```
* if I after this *check* the ACLs of `test-1` (or try to read again) that works perfectly fine

It seems to me that there is some form of asynchronous problem with which `setStreamMetadataRaw` resolves *before* the metadata are actually set.

**Comment by jageall** [comment]
There was a change made in v5 that ensures that acks are only returned after writes have been indexed by a majority of nodes. Can you confirm if this is an issue in v5.

**Comment by jonasholtkamp** [comment]
Tested it a little more in-depth now with ES versions 4.1.1, 4.1.2, 4.1.3-hotfix1 and 5.0.1. The initial ES version I opened this issue with was 4.1.1.

The median rate at which the `AccessDeniedError` occurs is about 10 times lower for 4.1.3-hotfix1 and 5.0.1. The *median* occurence for 4.1.3-hotfix1 and 5.0.1 is about 1000 runs, whereas it's 50 to 100 for 4.1.1 and 4.1.2.

I can't confirm that this issue is gone, however it occurs a lot less. We're currently running 4.1.3-hotfix1 in production so that helps a bit.

PS: Sorry for my late response, was on vacation.
